stdlib/public/Atomics/AtomicIntegers.swift.gyb

```diff
@@ -0,0 +1,275 @@
+      ("Int",    "Word",  "_builtinWordValue"),
+      ("UInt",   "Word",  "_builtinWordValue"),
+      ("Int64",  "Int64", "_value"),
+      ("UInt64", "Int64", "_value"),
+      ("Int32",  "Int32", "_value"),
+      ("UInt32", "Int32", "_value"),
+      ("Int16",  "Int16", "_value"),
+      ("UInt16", "Int16", "_value"),
+      ("Int8",   "Int8",  "_value"),
+      ("UInt8",  "Int8",  "_value"),
+    ]
+
+  def rmwOrderings():
+    return [
+      # Swift enum case,         API name,                 documentation name,        llvm name,   failure name
+      ('relaxed',                'Relaxed',                'relaxed',                 'monotonic', 'monotonic'),
+      ('acquiring',              'Acquiring',              'acquiring',               'acquire',   'acquire'),
+      ('releasing',              'Releasing',              'releasing',               'release',   'monotonic'),
+      ('acquiringAndReleasing',  'AcquiringAndReleasing',  'acquiring-and-releasing', 'acqrel',    'acquire'),
+      ('sequentiallyConsistent', 'Serialized',             'sequentially consistent', 'seqcst',    'seqcst'),
+    ]
+
+  def loadOrderings():
+    return [
+      # Swift enum case,         API name,     doc name,                  llvm name
+      ('relaxed',                'Relaxed',    'relaxed',                 'monotonic'),
+      ('acquiring',              'Acquiring',  'acquiring',               'acquire'),
+      ('sequentiallyConsistent', 'Serialized', 'sequentially consistent', 'seqcst'),
+    ]
+
+  def storeOrderings():
+    return [
+      # Swift enum case,         API name,     doc name,                  llvm name
+      ('relaxed',                'Relaxed',    'relaxed',                 'monotonic'),
+      ('releasing',              'Releasing',  'releasing',               'release'),
+      ('sequentiallyConsistent', 'Serialized', 'sequentially consistent', 'seqcst'),
+    ]
+
+  def caseStatementForOrdering(ordering):
+    if ordering == 'sequentiallyConsistent':
+      return "default"
+    else:
+      return "case .{}".format(ordering)
+
+  # FIXME: Swift should provide intrinsics for arbitrary ordering pairs
+  def llvmOrders(rmw, load): # See llvm/Support/AtomicOrdering.h
+    def max(rmw, load):
+      if load == "acquire":
+        if rmw == "monotonic":
+          return "acquire"
+        if rmw == "release":
+          return "acqrel"
+      if load == "seqcst":
+        return "seqcst"
+      return rmw
+    return max(rmw, load) + "_" + load
+
+  def lowerFirst(str):
+    return str[:1].lower() + str[1:] if str else ""
+
+  def argLabel(label):
+    return label + ": " if label <> "_" else ""
+
+  integerOperations = [
+    # Swift name,         llvm name,  operator, label,  doc
+    ('WrappingIncrement', 'add',      '&+',     "by",   "wrapping add"),
+    ('WrappingDecrement', 'sub',      '&-',     "by",   "wrapping subtract"),
+    ('BitwiseAnd',        'and',      '&',      "with", "bitwise and"),
```

Comment:
Why the change to "bitwise and with," etc.?

Reply:
The label seems important to clarify the argument’s role in the `atomicLoadThenBitwiseAnd(at:with:ordering:)` variant, and it seems appropriate to apply it to this shorter version, too.

Reply:
Have you considered simply reversing the arguments?

```
atomicLoadThenWrappingIncrement(by:at:ordering:)
atomicLoadThenBitwiseAnd(_:at:ordering:)
```

(Incidentally, _TSPL_ calls it "overflow add" instead of "wrapping add"; I do think "wrapping" is more self-explanatory here though.)

Reply:
The order matters; it specifies whether the new or old value is returned.

Reply:
Well, the order is encoded in the name; the argument order doesn't really indicate anything. I like Xiaodi's suggestion, but I think I'd prefer to have the `with:` label for bitwise operations even in the usual case where the address is implicit in `self`:

```
foo.loadThenWrappingIncrement(by: 3, ordering: .relaxed)
bar.loadThenBitwiseAnd(with: baz, ordering: .acquiring)

Int.loadThenBitwiseOr(with: 27, at: ptr, ordering: .releasing)
```

Reply:
Oops, I missed that the comment was only about changing argument labels, not the full name. My mistake!